### DISCUSSION TITLE
Is it possible to change listening ports?

### DISCUSSION THREAD

**Discussion by yorek** [Azure/data-api-builder#1425] [Q&A] [answered]
Is it possible to set the listening TCP ports for service (instead of defaults 5000/5001) ?

**Comment by yorek** [comment, marked as answer]
While there are no specific configuration options in Data API builder at the moment, since it is a regular .NET Core application, you can use the `ASPNETCORE_URLS` variable to set the desired listening address and ports, as explained here:

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-6.0#url-prefixes-1

With PowerShell, for example, you can do something like:

```powershell
$env:ASPNETCORE_URLS="http://localhost:8080;https://localhost:8081"
```

if you are on Linux, instead:

```bash
export ASPNETCORE_URLS="http://localhost:8080;https://localhost:8081"
```

and the start Data API builder as usual via 

```shell
dab start
```

and it will be listening on port 8080 and 8081

**Reply by fgbustoa** [reply]
OK. 

So this means that i have an unique listening ports couple per machine. 

Thus i cannot expose more than one database per machine: if i prepare two config json files, for two different databases, only the first service that i start will succeed, since the second will try to start listening onto the same ports.... and will fail.

**Reply by yorek** [reply]
That's a good point. I would say that right now, if you need to run DAB on the same machine, the best option is to run it in a container, so that each container can listen to a different port. For example:

```shell
docker run --name dab1 -it -v $pwd\config:/App/config -p 8080:5000 mcr.microsoft.com/azure-databases/data-api-builder:latest --ConfigFileName ./config/dab-config.json
```

and the you can run another one mapping it to port 8090:

```shell
docker run --name dab2 -it -v $pwd\config:/App/config -p 8090:5000 mcr.microsoft.com/azure-databases/data-api-builder:latest --ConfigFileName ./config/dab-config.json
```

provided that your config file is in the `config` folder.

I would recommend using docker-compose if also your database is running in a container, so that to simplify networking configuration:  https://github.com/Azure/data-api-builder/blob/main/docs/running-using-a-container.md

**Reply by ClaudioG64** [reply]
Thank you @yorek (Ciao Davide !), great suggestions.

If you run run from Powershell something like:
```
$env:ASPNETCORE_URLS="http://$($env:ComputerName):8080;https://$($env:ComputerName):8081"
dab start
```
You end with a network accessible endpoint on all of your IPv4/v6 addresses.

So if you want to host multiple "Db-Api" pairs on a single machine, you just have to run multiple instances of Powershell with different ASPNETCORE_URLS env variable for each process.

**Reply by yorek** [reply]
Good point, thanks @ClaudioG64! (So happy to see you around!!!)

**Comment by seantleonard** [comment]
You can also modify the appsetting.json file with custom Kestrel endpoint settings to override the default ports per https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-6.0#configureiconfiguration
```json
{
  "Logging": {
    "LogLevel": {
      "Default": "Error",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedHosts": "*",
  "Kestrel": {
    "Endpoints": {
      "Https": {
        "Url": "https://localhost:6543"
      }
    }
  }
}
```

**Reply by acolville** [reply]
I want to change to an IP address, rather than localhost to allow connections across the LAN, but neither of these options work. DAB always starts listening on localhost:5000/5001. Is it possible to change this?

**Reply by seantleonard** [reply]
How are you launching DAB? If your are using VSCode, you may need change the launchSettings.json file.

**Comment by acolville** [comment]
From a terminal window, I will try and find the LaunchSettings.json file is it in System32?


Alistair Colville
Director

***@***.***
t:
e:
w:
01983 882874
***@***.******@***.***>
https://www.colvillecare.co.uk<https://www.colvillecare.co.uk/>

**Reply by seantleonard** [reply]
Do you use `dotnet run` to launch DAB from the terminal? System32 should not be touched for changing DAB configuration. Here are the two files that are applicable to port/ip in the DAB project: 
- https://github.com/Azure/data-api-builder/blob/main/src/Service/Properties/launchSettings.json
- https://github.com/Azure/data-api-builder/blob/main/src/Service/appsettings.json

In addition, here is the documentation for various hosting options of asp.net projects (which DAB is an asp.net project). You may benefit from hosting DAB with one of these mechanisms which their own specific ways of configuring host settings:
- https://learn.microsoft.com/aspnet/core/host-and-deploy/?view=aspnetcore-6.0

**Comment by acolville** [comment]
I followed the example for on-prem
Overview to Azure data API builder | Microsoft Learn<https://learn.microsoft.com/en-us/azure/data-api-builder/overview-to-data-api-builder?tabs=azure-sql#run-data-api-builder-on-premises>

so to run I just type” dab start”

I will see if I can track down those files and have a look.

Alistair Colville
Director

***@***.***
t:
e:
w:
01983 882874
***@***.******@***.***>
https://www.colvillecare.co.uk<https://www.colvillecare.co.uk/>

**Comment by acolville** [comment]
Many thanks, that is all running now.

Alistair Colville
Director

***@***.***
t:
e:
w:
01983 882874
***@***.******@***.***>
https://www.colvillecare.co.uk<https://www.colvillecare.co.uk/>

**Comment by pholcroft** [comment]
I just wanted to contribute another method to the above using a **.env** file, although you lose the extra security of an environment variable.
You only need a certificate if using https.

C:\ContactsAPI\\.env
C:\ContactsAPI\dab-config.json
C:\ContactsAPI\Certificate\ContactsAPI.pfx


**.env**
АЅРΝЕТСОRЕ_URLЅ=httрѕ://*:5090
ASPNETCORE_Kestrel__Certificates__Default__Password="REPLACEWITHARANDOMPASSWORD" 
ASPNETCORE_Kestrel__Certificates__Default__Path=C:\ContactsAPI\Certificate\ContactsAPI.pfx

**CreateCert.ps1**
$cert = New-SelfSignedCertificate -DnsName @("localhost") -CertStoreLocation "cert:\LocalMachine\My" -NotAfter (Get-Date).AddYears(10)
$certKeyPath = "C:\ContactsAPI\Certificate\ContactsAPI.pfx"
$password = ConvertTo-SecureString 'REPLACEWITHARANDOMPASSWORD' -AsPlainText -Force
$cert | Export-PfxCertificate -FilePath $certKeyPath -Password $password
